firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/MutationSquash.java

```diff
@@ -0,0 +1,119 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.model.mutation;
+
+import static com.google.firebase.firestore.util.Assert.fail;
+import static com.google.firebase.firestore.util.Assert.hardAssert;
+
+import com.google.common.collect.Sets;
+import com.google.firebase.firestore.model.FieldPath;
+import com.google.firebase.firestore.model.MutableDocument;
+import com.google.firebase.firestore.model.ObjectValue;
+import com.google.firestore.v1.Value;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Utility class to compute a mutation from an original document, and a mutated version. There are
+ * likely multiple mutations happened in between, hence "squash".
+ */
+public class MutationSquash {
+  public enum Type {
+    Set,
+    Delete,
+    Patch,
+    // Patch but with precondition set to None.
+    PatchWithMerge,
+    None
+  }
+
+  /**
+   * Give an original document, its mutated version, and the squash target type, returns a squashed
+   * mutation, that would hold invariant: original + squashed_mutation = finalDoc.
+   */
+  @Nullable
+  public static Mutation toMutation(MutableDocument original, MutableDocument finalDoc, Type type) {
+    switch (type) {
+      case Set:
+        return toSetMutation(original, finalDoc);
+      case Delete:
+        return toDeleteMutation(original, finalDoc);
+      case Patch:
+        return toPatchMutation(original, finalDoc, false);
+      case PatchWithMerge:
+        return toPatchMutation(original, finalDoc, true);
+      case None:
+        return null;
+      default:
+        fail("Unrecognized squash mutation type");
+    }
+    return null;
+  }
+
+  private static SetMutation toSetMutation(MutableDocument original, MutableDocument finalDoc) {
+    hardAssert(
```

Comment:
You can move this assert to the callsite and thus share the code for all methods.